### ISSUE TITLE
A couple things about new "thread-safary" code of CCFileUtils.cpp, etc

### ISSUE THREAD

**Issue by gestern** [cocos2d/cocos2d-x#19059]
First of all I can't figure out the purpose for which these changes were made? What is the purpose of capturing a recursive mutex for further reading the values of class member variables and capturing for writing? For example: https://github.com/cocos2d/cocos2d-x/blob/f8ed20e4b3bf74f0812e019d58d823d72ef8d4a9/cocos/platform/CCFileUtils.cpp#L922-L926

Why is it now impossible to access I/O operations without synchronization at the level of recursive heavyweight mutexes, even if reading and writing come from the same thread? In an attempt to exclude a potential deadlock while I/O, a recursive mutex is used instead of a thoughtful architecture. Finally, a cherry on the cake is a macro `DECLARE_GUARD`. Now my regression tests have noticed a performance drop of -15% for I/O operations. I would like to understand why this was done?

**Comment by crazyhappygame** [comment]
1. Could you share your regressions tests?
2. Please grep for performOperationOffthread - basically each file operation can be done in async way. Before there was race condition if some one call async (new thread) and sync (main thread) operation
3. I agree that recursive_mutex should not be used at all. Probably better would be to just use std::mutex when needed. But that will require to re-design of FileUtils.

**Comment by crazyhappygame** [comment]
4. Most probably instead of mutex it should be possible to reorganize code in that way that only I/O (read/write file) operations are run in separate thread and other operation e.g. read/write  _searchResolutionsOrderArray are done in main thread

**Comment by gestern** [comment]
First of all I want to describe a typical scenario for the game. The introductory data are such that I proceed from the fact that the **average duration of a gaming session is approximately 15 minutes**. During this time, the typical behavior is that if the game has any analytics uploaded to the server mediation data as well as data on the game events (the state of sound and music, etc.), then for 15 minutes of the game session it runs up to 40-50 calls to the system data input-output to the file system.

```c++
void performTest(int count) {
    for (int i = 0; i < count; i++) {
        auto fu = FileUtils::getInstance();
        std::string testDir = fu->getWritablePath() + "/test" + std::to_string(i);
        fu->createDirectory(testDir);
        
        std::string fileName = testDir + "/somefile.txt";
        {
            ValueVector vv;
            for(int x = 0 ; x < 10 ; x++) {
                vv.push_back(Value(x));
            }
            
            CCASSERT(fu->writeValueVectorToFile(vv, fileName), "My master thinks we are losing this war");
        }
        
        fu->addSearchPath(testDir);
        
        {
            ValueVector vv = fu->getValueVectorFromFile(fileName);
            CCASSERT(vv.size() == 10, "My master informs you: he cannot help you");
        }
        
        if(fu->isDirectoryExist(testDir)) {
            fu->removeDirectory(testDir);
        }
        
        
        
        fu->setSearchPaths({""});
    }
}

// within bool HelloWorld::init()
//..
auto button = ui::Button::create("HelloWorld.png");
    button->setPosition(visibleSize/2);
    button->addClickEventListener([&](Ref* source){
        FileUtils::destroyInstance();
        ProfilingBeginTimingBlock("files-test");
        
        performTest(100);
        
        ProfilingEndTimingBlock("files-test");
        Profiler::getInstance()->displayTimers();
        
    });
    addChild(button);
//..
```
The following code can give a possible output:
<img width="332" alt="image" src="https://user-images.githubusercontent.com/1861589/45841811-95d3f080-bd23-11e8-88f7-894ef62c2b34.png">
Closing eyes to OS X data, which have gotten on OSX (Quad Core CPU). Here, too many external factors can influence the integrity of the measurements, so in this case it is necessary to increase the duration of the session (thus increasing the magnitude of the measurement error). And as I said in start topic, right now I can safely say that the added mutexes stably negatively affect the performance in all cases in all devices.

So, as we well remember from the course of [computer science](https://en.wikipedia.org/wiki/Inter-process_communication) File object is a object of IPC. My opinion on this code is this: wrapping operations with files with operations with mutexes identically equals the situation when mutexes are wrapped over other mutexes.

**Comment by crazyhappygame** [comment]
@minggo @PatriceJiang FileUtils std::recursive_mutex impact significantly performance of I/O operations 15% -26%. Could you fix it?
I think that we should remove FileUtils mutex at all and reorganize FileUtils in that way that update of all FileUtils members is done in main thread and performOperationOffthread are only I/O operation (e.g. file read/write).

**Comment by minggo** [comment]
@gestern the purpose of the codes is to make FileUtils thread safe as it may be used in other thread. As the performance is critical for cocos, so i think 15% performance dropped is not acceptable. So there are two choices:

* use mutex as needed
* revert the PR, developers should do thread safe related file operation themselves

**Comment by gestern** [comment]
There are a lot of solutions here by adding "more security" (I say in quotes because I take it as a rule not to use significant IO operations during render cycles. Thus, due to proper design, I avoid the problem with a possible race in these operations) without dropping the performance via applying architectural solutions:

1.  To be able to apply a strategy design pattern to `FileUtils` class and use mutex in demand cases `FileUtils::getInstance(StrategyLocker locker = StrategyLocker::NoLock by default or StrategyLocker::Mutex)` or `FileUtils::getInstance<Locker>()` this is not important and representation can be invented during development;
2.  or any significant addition should be made as being disabled by using macros (for example: `FILE_API_SYNCED`) and later wrapped al mutextes around `#ifdef ... #endif` for example: 
  ```cpp
  ...  getValueVectorFromFile() {
  #ifdef FILE_API_SYNCED
  // mutex locker
  #endif
  .. operations
  ```
3. or if you want to do analogy with POSIX (required by IEEE Std 1003.1-2001) to make analogues thread-safe call to each none-thread-safe call, like by [adding with suffix](http://pubs.opengroup.org/onlinepubs/000095399/xrat/xsh_chap02.html#tag_03_02_09_08) `_r`;
4. or, for example, guided by solutions by well [base on known books of Herb Sutter](https://www.infoq.com/news/2014/10/cpp-lock-free-programming) to avoid hard synchronization on reading, add synchronization without a huge performance subsidence

**Comment by crazyhappygame** [comment]
@minggo @PatriceJiang I checked few async call from FileUtils and for me it seems like some of them does not need lock at all (like example below). Other async call can be refactored in that way that all FileUtils members updates are done in main thread and only I/O operation is in separate thread
```
void FileUtils::writeStringToFile(std::string dataStr, const std::string& fullPath, std::function<void(bool)> callback)
{
    performOperationOffthread([fullPath](const std::string& dataStrIn) -> bool {
        return FileUtils::getInstance()->writeStringToFile(dataStrIn, fullPath);
    }, std::move(callback),std::move(dataStr));
}
```

**Comment by PatriceJiang** [comment]
I have executed the use case several times in a row, and I found that even when the same code is executed multiple times, the output can be very different.

Below is a plot of the `avg2` data generated on the IOS. Whether it's locked or unlocked, you'll find that the results are overlapping. (I also got similar results on the MAC.)

![ios_all](https://user-images.githubusercontent.com/40414978/46052019-f3769b80-c16e-11e8-8da9-d57effd9a201.png)


I think the test code cannot be used to measure the performance of `FileUtils`.

In fact, **`locks` are quite fast compared to filesystem operations**. I have tested `std::recursive_mutex` on my PC,  for a full `lock&unlock` that only takes less than `100ns`. **This cost is minimal compared to file operations.**

**Comment by gestern** [comment]
>I have executed the use case several times in a row, and I found that even when the same code is executed multiple times, the output can be very different.

Ready to suggest that this happens because you are making dependent measurements, influenced by multiple factors, for example [Page replacement algorithm](https://en.wikipedia.org/wiki/Page_replacement_algorithm) or/and [Branch predictor](https://en.wikipedia.org/wiki/Branch_predictor). I am now talking about the volatility of the chart relative to the x-coord. But without seeing the final code (let's say it's not clear what the difference is in `lock1` and `lock2`, `lock3`)

I want to add that I also continued theoretical research on the extent to which your new code with mutexes can affect overall performance. And I came to the conclusion that, under certain conditions, the code can significantly reduce the speed. What are the conditions? For example frequent use of TTF-fonts with different sizes of glyphs or avoid to use atlases with textures. Each `Label` will invoke `initWithTTFConfig()` which invoke `isFileExist()`.   The root of the problem in `isFileExist()` which calls the blocking call to `fullPathForFilename()`: https://github.com/cocos2d/cocos2d-x/blob/aca3016bee4db8d9f2edf57a8519b12ef75256d5/cocos/platform/CCFileUtils.cpp#L815-L864

By the way, this can exacerbate an already existing problem when downloading games on platforms (Android), where access to the file system is especially costly and making an additional lock only aggravate the problem.
